### DISCUSSION TITLE
Ideas how to improve the performance

### DISCUSSION THREAD

**Discussion by szaimen** [nextcloud/all-in-one#5254] [Ideas] [answered]
All ideas that improve the performance of AIO are always welcome. Feel free to add them below.

See https://github.com/nextcloud/all-in-one/discussions/1335 for general performance recommendations.

**Comment by szaimen** [comment, marked as answer]
I've added this to the list in https://github.com/nextcloud/all-in-one/issues/5251 now but feel free to add further ideas below.

**Comment by jasonmwade** [comment]
I would suggest the first thing to be able to improve the performance is some way to view the current performance and change the features of the software based on the hardware. 

Suggest creating a method to do a performance evaluation of the current setup testing CPU, DISK, Network, memory, etc and make recommendations like disabling image thumbnails or increasing database memory. Then actually getting NC to make those changes.

Performance roll-ups like number of slow queries per day, time spent generating thumbnails, add-ons causing delays, external call timeouts, etc. 

Create t-shirt configurations for low-spec, mid-spec, and high-spec hardware that reconfigure based on the preferences of the admin or user

Categorize add-ons based on their performance impact.

**Reply by szaimen** [reply]
Hi, thanks for the idea! I agree with you in general that it would help to have some kind of baseline to compare performance changes with. However since the project is very complex and consists of many components, I guess it would be more helpful to do some change and then think about how to do a before - after comparision in terms of performance for this specific change as anything else probably too complicated.

**Reply by szaimen** [reply]
Do you have some more fine grained ideas how to improve the performance  in detail maybe?

**Reply by jasonmwade** [reply]
I use the Linuxserver.io image because the AIO install requires ports to be opened to install (https://github.com/nextcloud/all-in-one?tab=readme-ov-file#can-i-run-aio-offline-or-in-an-airgapped-system). 
I have 3 different Nextcloud installs right now locally which seems impossible with AIO since they would use the same ports. I run on a Proxmox cluster, so I can change the performance by updating the VM spec or moving the VM to a different physical machine. 

Since I cannot run AIO because I'm not willing to open ports directly from the internet to Nextcloud, I can only offer suggestions generally.

Specific Recommendation 1: Write a section in the readme.md for performance

There's no mention of performance in the main readme.md other than 3 sentences about slow storage options. There's no description about hardware recommendations. The hardware the instance is running on makes a drastic change in performance. Slow disks, low amounts of ram, and other noticeable degraded performance are unavoidable on very low spec hardware. No amount of tweaking is going to overcome hardware limitations. However, this doesn't seem outlined at all in the project. 

Specific Recommendation 2: Have some way to adjust common performance settings from the interface.

There exist on the internet helpful guides to adjust php-fpm settings, tweak memory, changing database ram, but these all require dropping to a cli and updating files critical to the operation of Nextcloud. This could be variables in the docker run command, changes to the configuration files, etc.

**Reply by szaimen** [reply]
Hi, thanks for the ideas. 

Regarding 1.: We have this article: https://github.com/nextcloud/all-in-one/discussions/1335. Should we add some more details regarding hardware to it?

**Comment by PinchePelon** [comment]
Add environment variables to allow changing the amount of process manager servers that PHP-FPM can use. (pm.max_children, pm.start_servers, pm.min/max_spare_servers). The default values are pretty low, and can cause performance issues over time.

**Reply by szaimen** [reply]
Hi, thanks for the idea! Are you actually using Nextcloud AIO or something else like e.g. the community docker container https://hub.docker.com/_/nextcloud ? 

And if you should be using AIO, did you run into issues with AIOs php-fpm config?
